stats/opencensus/e2e_test.go

```diff
@@ -0,0 +1,773 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package opencensus
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/google/go-cmp/cmp"
+	"go.opencensus.io/stats/view"
+	"go.opencensus.io/tag"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/internal/grpctest"
+	"google.golang.org/grpc/internal/leakcheck"
+	"google.golang.org/grpc/internal/stubserver"
+	"google.golang.org/grpc/test/grpc_testing"
+)
+
+type s struct {
+	grpctest.Tester
+}
+
+func Test(t *testing.T) {
+	grpctest.RunSubTests(t, s{})
+}
+
+func init() {
+	// OpenCensus, once included in binary, will spawn a global goroutine
+	// recorder that is not controllable by application.
+	// https://github.com/census-instrumentation/opencensus-go/issues/1191
+	leakcheck.RegisterIgnoreGoroutine("go.opencensus.io/stats/view.(*worker).start")
+}
+
+var defaultTestTimeout = 5 * time.Second
+
+type fakeExporter struct {
+	t *testing.T
+
+	mu        sync.RWMutex
+	seenViews map[string]*viewInformation
+}
+
+// viewInformation is information Exported from the view package through
+// ExportView relevant to testing, i.e. a reasonably non flaky expectation of
+// desired emissions to Exporter.
+type viewInformation struct {
+	aggType    view.AggType
+	aggBuckets []float64
+	desc       string
+	tagKeys    []tag.Key
+	rows       []*view.Row
+}
+
+func (fe *fakeExporter) ExportView(vd *view.Data) {
+	fe.mu.Lock()
+	defer fe.mu.Unlock()
+	fe.seenViews[vd.View.Name] = &viewInformation{
+		aggType:    vd.View.Aggregation.Type,
+		aggBuckets: vd.View.Aggregation.Buckets,
+		desc:       vd.View.Description,
+		tagKeys:    vd.View.TagKeys,
+		rows:       vd.Rows,
+	}
+}
+
+func (vi *viewInformation) equal(vi2 *viewInformation) bool {
+	if vi == nil && vi2 == nil {
+		return true
+	}
+	if (vi != nil) != (vi2 != nil) {
+		return false
+	}
+	if vi.aggType != vi2.aggType {
+		return false
+	}
+	if !cmp.Equal(vi.aggBuckets, vi2.aggBuckets) {
+		return false
+	}
+	if vi.desc != vi2.desc {
+		return false
+	}
+	if !cmp.Equal(vi.tagKeys, vi2.tagKeys, cmp.Comparer(func(a tag.Key, b tag.Key) bool {
+		return a.Name() == b.Name()
+	})) {
+		return false
+	}
+	if !compareRows(vi.rows, vi2.rows) {
+		return false
+	}
+	return true
+}
+
+// compareRows compares rows with respect to the information desired to test.
+// Both the tags representing the rows and also the data of the row are tested
+// for equality. Rows are in nondeterministic order when ExportView is called,
+// but handled because ExportView is called every default metrics reporting
+// interval, which will work eventually in a reasonable manner (i.e. doesn't
+// flake).
+func compareRows(rows []*view.Row, rows2 []*view.Row) bool {
+	if rows == nil && rows2 == nil {
+		return true
+	}
+	if (rows != nil) != (rows2 != nil) {
+		return false
+	}
+	if len(rows) != len(rows2) {
```

Comment:
I think if you have this, you don't really need either of the two checks above.  If both have length zero, nothing will happen and we'll consider them equal.  (This means `nil` == `[]Row{}`, but if we need to be sensitive to a difference between those two then I'm a little worried.)